---
title: "Scatter plot tutorial æ•£ç‚¹å›¾æ•™ç¨‹"
author: "Defeng Bai(ç™½å¾·å‡¤), Chuang Ma(é©¬é—¯), Jiani Xun(è€ä½³å¦®)"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    theme: cerulean
    highlight: haddock
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: no
      smooth_scroll: yes
    code_fold: show
  word_document:
    toc: yes
    toc_depth: '3'
  pdf_document:
    toc: yes
    toc_depth: '3'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  collapse = T, echo=T, comment="#>", message=F, warning=F,
	fig.align="center", fig.width=5, fig.height=3, dpi=150)
```



The Scatter plot scripts is referenced from MicrobiomeStatPlot [Inerst Reference below].

If you use this script, please cited å¦‚æœä½ ä½¿ç”¨æœ¬ä»£ç ï¼Œè¯·å¼•ç”¨ï¼š

**Yong-Xin Liu**, Lei Chen, Tengfei Ma, Xiaofang Li, Maosheng Zheng, Xin Zhou, Liang Chen, Xubo Qian, Jiao Xi, Hongye Lu, Huiluo Cao, Xiaoya Ma, Bian Bian, Pengfan Zhang, Jiqiu Wu, Ren-You Gan, Baolei Jia, Linyang Sun, Zhicheng Ju, Yunyun Gao, **Tao Wen**, **Tong Chen**. 2023. EasyAmplicon: An easy-to-use, open-source, reproducible, and community-based pipeline for amplicon data analysis in microbiome research. **iMeta** 2(1): e83. https://doi.org/10.1002/imt2.83

The online version of this tuturial can be found in https://github.com/YongxinLiu/MicrobiomeStatPlot


**Authors**

First draft(åˆç¨¿)ï¼šDefeng Bai(ç™½å¾·å‡¤)ï¼›Proofreading(æ ¡å¯¹)ï¼šMa Chuang(é©¬é—¯) and Jiani Xun(è€ä½³å¦®)ï¼›Text tutorial(æ–‡å­—æ•™ç¨‹)ï¼šDefeng Bai(ç™½å¾·å‡¤)


# Introductionç®€ä»‹

## Scatter plot æ•£ç‚¹å›¾
ä»€ä¹ˆæ˜¯æ•£ç‚¹å›¾ï¼Ÿ
What is a scatter plot?

æ•£ç‚¹å›¾ï¼ˆScatter Plotï¼‰ æ˜¯ä¸€ç§ç»Ÿè®¡å›¾å½¢ï¼Œç”¨äºå±•ç¤ºä¸¤ä¸ªå˜é‡ä¹‹é—´çš„å…³ç³»ã€‚å®ƒé€šè¿‡äºŒç»´åæ ‡ç³»ä¸­çš„ç‚¹æ¥è¡¨ç¤ºè§‚å¯Ÿæ•°æ®ï¼Œæ¯ä¸ªç‚¹çš„ä½ç½®ç”±ä¸¤ä¸ªå˜é‡çš„å–å€¼å†³å®šï¼Œæ¨ªåæ ‡è¡¨ç¤ºè‡ªå˜é‡ï¼ˆX è½´ï¼‰ï¼Œçºµåæ ‡è¡¨ç¤ºå› å˜é‡ï¼ˆY è½´ï¼‰ã€‚æ•£ç‚¹å›¾çš„åº”ç”¨ï¼šç›¸å…³æ€§åˆ†æ,åˆ¤æ–­ä¸¤ä¸ªå˜é‡ä¹‹é—´æ˜¯å¦å­˜åœ¨æŸç§å…³è”ï¼Œä»¥åŠè¿™ç§å…³è”æ˜¯æ­£ç›¸å…³ã€è´Ÿç›¸å…³è¿˜æ˜¯æ— å…³ã€‚æ¨¡å‹æ‹Ÿåˆ,é€šè¿‡è§‚å¯Ÿæ•£ç‚¹å›¾ï¼Œå¯ä»¥é€‰æ‹©åˆé€‚çš„å›å½’æ¨¡å‹ï¼ˆå¦‚çº¿æ€§å›å½’ã€éçº¿æ€§å›å½’ç­‰ï¼‰æ¥æ‹Ÿåˆæ•°æ®ã€‚å¼‚å¸¸æ£€æµ‹,å¯ä»¥é€šè¿‡è§‚å¯Ÿæ•£ç‚¹å›¾ä¸­çš„å¼‚å¸¸ç‚¹æ¥æ£€æµ‹æ½œåœ¨çš„å¼‚å¸¸æ•°æ®æˆ–é”™è¯¯è§‚æµ‹å€¼ã€‚

A Scatter Plot is a statistical graph used to show the relationship between two variables. It represents the observed data by means of points in a two-dimensional coordinate system, where the position of each point is determined by the values of the two variables, with the horizontal coordinate representing the independent variable (X-axis) and the vertical coordinate representing the dependent variable (Y-axis). Scatterplot application: Correlation analysis, to determine whether there is some kind of correlation between two variables, and whether the correlation is positive, negative or irrelevant. Model fitting, by looking at the scatterplot, you can choose a suitable regression model (e.g. linear regression, non-linear regression, etc.) to fit the data. Anomaly detection, by looking at the anomalies in the scatterplot, you can detect potential anomalies or erroneous observations.


å…³é”®å­—ï¼šå¾®ç”Ÿç‰©ç»„æ•°æ®åˆ†æã€MicrobiomeStatPlotã€æ•£ç‚¹å›¾ã€Rè¯­è¨€å¯è§†åŒ–

Keywords: Microbiome analysis, MicrobiomeStatPlot, Scatter plot, R visulization



## Scatter plot examples æ•£ç‚¹å›¾åº”ç”¨æ¡ˆä¾‹

### Example1 æ¡ˆä¾‹1

è¿™æ˜¯Patrick Lehodeyåœ¨2020å¹´å‘è¡¨äºBiogeosciencesä¸Šçš„æ–‡ç« ï¼Œç¬¬ä¸€ä½œè€…ä¸ºAudrey Delpechï¼Œé¢˜ç›®ä¸ºï¼šInfluence of oceanic conditions in the energy transfer efficiency estimation of a micronekton model.  https://bg.copernicus.org/articles/17/833/2020/

This is an article by Patrick Lehodey published in Biogeosciences in 2020, first authored by Audrey Delpech, titled: Influence of oceanic conditions in the energy transfer efficiency estimation of a micronekton model. https://bg.copernicus.org/articles/17/833/2020/


![](eg1.png)

Figure 5 Scatter plot and marginal distribution from kernel density estimation in the plane (a) (ğ’¯,ğ’±) and (b) (ğ’¯,ğ’®) for the configurations corresponding to Exp. 3a, b, c and d from Table 3.

å›¾5 æ˜¾ç¤ºäº†åœ¨å¹³é¢ (a) (ğ’¯,ğ’±)å’Œ(b)(ğ’¯,ğ’®)ä¸ŠåŸºäºæ ¸å¯†åº¦ä¼°è®¡çš„æ•£ç‚¹å›¾åŠè¾¹ç¼˜åˆ†å¸ƒï¼Œåˆ†åˆ«å¯¹åº”è¡¨3ä¸­å®éªŒ3aã€3bã€3c å’Œ 3d çš„é…ç½®ã€‚


**ç»“æœ**ï¼š

Figure 5 shows that the distributions along the secondary variables of each configuration are close enough for the experiments to be compared, avoiding any risk of cross-correlation. 

å›¾ 5 æ˜¾ç¤ºï¼Œæ¯ç§é…ç½®çš„æ¬¡è¦å˜é‡åˆ†å¸ƒéƒ½éå¸¸æ¥è¿‘ï¼Œè¶³ä»¥è¿›è¡Œå®éªŒæ¯”è¾ƒï¼Œé¿å…äº†ä»»ä½•äº¤å‰ç›¸å…³çš„é£é™©ã€‚



### Example2 æ¡ˆä¾‹2

è¿™æ˜¯Masahira Hattori åœ¨2022å¹´å‘è¡¨äºNCä¸Šçš„æ–‡ç« ï¼Œç¬¬ä¸€ä½œè€…ä¸ºSuguru Nishijimaï¼Œé¢˜ç›®ä¸ºï¼šExtensive gut virome variation and its associations with host and environmental factors in a population-level cohort  https://doi.org/10.1038/s41467-022-32832-W

This is an article by Patrick Lehodey published in Biogeosciences in 2020, first authored by Audrey Delpech, titled: Influence of oceanic conditions in the energy transfer efficiency estimation of a micronekton model

![](eg2.png)
Fig. 1: Overview of reconstructed phage genomes from 4198 human gut metagenomes.

å›¾ 1ï¼šä» 4198 ä¸ªäººç±»è‚ é“å…ƒåŸºå› ç»„ä¸­é‡å»ºçš„å™¬èŒä½“åŸºå› ç»„æ¦‚è§ˆ

a. Genome size and GC content of phage genomes (n = 4709) reconstructed from the dataset of 4,198 whole metagenomes. Bar plots on the top and right side depict the distribution of genome size and GC content, respectively. 

a. ä» 4198 ä¸ªå…¨å…ƒåŸºå› ç»„æ•°æ®é›†ä¸­é‡å»ºçš„å™¬èŒä½“åŸºå› ç»„ï¼ˆn = 4709ï¼‰çš„åŸºå› ç»„å¤§å°å’Œ GC å«é‡ã€‚é¡¶éƒ¨å’Œå³ä¾§çš„æŸ±çŠ¶å›¾åˆ†åˆ«æè¿°äº†åŸºå› ç»„å¤§å°å’Œ GC å«é‡çš„åˆ†å¸ƒæƒ…å†µã€‚


**ç»“æœ**ï¼š

Clustering of the 4709 phage sequences with >95% sequence similarity generated 1347 viral operational taxonomic units (vOTUs) (corresponding to the species level) (Fig. 1a and Supplementary Fig. 2a, Supplementary Data 4).
å¯¹åºåˆ—ç›¸ä¼¼åº¦å¤§äº 95% çš„ 4709æ¡å™¬èŒä½“åºåˆ—è¿›è¡Œèšç±»ï¼Œç”Ÿæˆäº†1347ä¸ªç—…æ¯’æ“ä½œåˆ†ç±»å•å…ƒï¼ˆvOTUsï¼‰ï¼ˆç›¸å½“äºç‰©ç§æ°´å¹³ï¼‰ï¼ˆå›¾ 1a å’Œè¡¥å……å›¾ 2aï¼Œè¡¥å……æ•°æ® 4ï¼‰ã€‚


### Example3 æ¡ˆä¾‹3

è¿™æ˜¯Damian R. plichta åœ¨2023å¹´å‘è¡¨äºNature Microbiologyä¸Šçš„æ–‡ç« ï¼Œç¬¬ä¸€ä½œè€…ä¸ºJoachim Johansen
ï¼Œé¢˜ç›®ä¸ºï¼šCentenarians have a diverse gut virome with the potential to modulate metabolism and promote healthy lifespan  https://www.nature.com/articles/s41467-024-45793-z

This is an article by Damian R. plichta published in Nature Microbiology in 2023, first authored by Joachim Johansen with the title : Centenarians have a diverse gut virome with the potential to modulate metabolism and promote healthy lifespan. https://www.nature.com/articles/s41467-024-45793-z

![](eg3.png)

Fig. 3 | Phage signatures correlate with the unique centenarian bacterial 
communities.
å›¾ 3 | å™¬èŒä½“ç‰¹å¾ä¸ç‹¬ç‰¹çš„ç™¾å²è€äººç»†èŒç¾¤è½ç›¸å…³ç¾¤è½ç›¸å…³ã€‚

d, Several enriched temperate vOTUs displayed PtW abundance (log10 scale) trends with MSP host abundance, such as the abundance of E. boltea vOTUs showing a Spearman rank correlation of 0.51 (P = 4.82 Ã— 10âˆ’66), C. scindens (P = 4.49 Ã— 10âˆ’08) and P. distasonis (P = 6.00 Ã— 10âˆ’11). 
d, ä¸€äº›å¯Œé›†çš„æ¸©å’Œ vOTUs çš„ PtW ä¸°åº¦ï¼ˆlog10 æ ‡åº¦ï¼‰ä¸ MSP å®¿ä¸»ä¸°åº¦å‘ˆç›¸å…³è¶‹åŠ¿ï¼Œå¦‚ E. boltea vOTUs ä¸°åº¦çš„ Spearman ç­‰çº§ç›¸å…³æ€§ä¸º 0.51ï¼ˆP = 4.82 Ã— 10-66ï¼‰ã€C. scindensï¼ˆP = 4.49 Ã— 10-08ï¼‰å’Œ P. distasonisï¼ˆP = 6.00 Ã— 10-11ï¼‰ã€‚

**ç»“æœ**ï¼š

We hypothesized that the trend could reflect the abundance profile of the viral bacterial hosts and thus correlated the viral profiles of predicted temperate viruses to their predicted host. We found that the profile of overabundant bacteria such as C. scindens and its associated viruses were significantly correlated (Cor = 0.30, P = 4.49 Ã— 10âˆ’08). A similar trend was found for Akkermansia muciniphila (Cor = 0.25, P = 6.23 Ã— 10âˆ’11), Enterocloster bolteae (Cor = 0.51, P = 4.82 Ã— 10âˆ’66) and Parabacteroides distasonis (Cor = 0.35, P = 6.00 Ã— 10âˆ’11) (Fig. 3d). 

æˆ‘ä»¬æ¨æµ‹è¿™ä¸€è¶‹åŠ¿å¯èƒ½åæ˜ äº†ç—…æ¯’ç»†èŒå®¿ä¸»çš„ä¸°åº¦ã€‚å› æ­¤ï¼Œæˆ‘ä»¬å°†é¢„æµ‹çš„æ¸©å¸¦ç—…æ¯’çš„ç—…æ¯’ç‰¹å¾ä¸å…¶é¢„æµ‹çš„å®¿ä¸»ç›¸å…³è”ã€‚é¢„æµ‹çš„æ¸©å¸¦ç—…æ¯’çš„ç—…æ¯’ç‰¹å¾ä¸å…¶é¢„æµ‹çš„å®¿ä¸»ç›¸å…³è”ã€‚æˆ‘ä»¬å‘ç°C.scindensç­‰è¿‡é‡ç»†èŒçš„ç‰¹å¾ä¸ç›¸å…³ç—…æ¯’æœ‰æ˜¾è‘—ç›¸å…³æ€§ï¼ˆCor = 0.30, P = 4.49 Ã— 10-08ï¼‰ã€‚åœ¨ Akkermansia muciniphilaä¸­ä¹Ÿå‘ç°äº†ç±»ä¼¼çš„è¶‹åŠ¿ï¼ˆCor = 0.25ã€ P=6.23Ã—10-11ï¼‰ã€Enterocloster bolteaeï¼ˆCor=0.51ï¼ŒP=4.82Ã—10-66ï¼‰å’Œ Parabacteroides distasonisï¼ˆCor = 0.35ï¼ŒP = 6.00 Ã— 10-11ï¼‰ï¼ˆå›¾ 3dï¼‰ã€‚

è¿™é‡Œä¸¾ä¾‹ç»˜åˆ¶æ•£ç‚¹å›¾çš„åŸºç¡€å›¾å½¢å’Œæ•£ç‚¹å›¾å’Œå…¶ä»–ç±»å‹å›¾çš„ç»„åˆ
Here we give an example of drawing the basic graphics of scatter plots and the combination of scatter plots and other types of graphs.


## Packages installationè½¯ä»¶åŒ…å®‰è£…

```{r}
# åŸºäºCRANå®‰è£…RåŒ…ï¼Œæ£€æµ‹æ²¡æœ‰åˆ™å®‰è£…
p_list = c("ggplot2", "ggpubr", "ggpmisc", "doBy", "FactoMineR", "factoextra", 
           "tidyverse", "ggExtra", "vegan", "cowplot", "MASS", "scales","showtext","grid")
for(p in p_list){if (!requireNamespace(p)){install.packages(p)}
    library(p, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)}

# åŠ è½½RåŒ… Load the package
suppressWarnings(suppressMessages(library(ggplot2)))
suppressWarnings(suppressMessages(library(ggpubr)))
suppressWarnings(suppressMessages(library(ggpmisc)))
suppressWarnings(suppressMessages(library(doBy)))
suppressWarnings(suppressMessages(library(FactoMineR)))
suppressWarnings(suppressMessages(library(factoextra)))
suppressWarnings(suppressMessages(library(tidyverse)))
suppressWarnings(suppressMessages(library(ggExtra)))
suppressWarnings(suppressMessages(library(vegan)))
suppressWarnings(suppressMessages(library(cowplot)))
suppressWarnings(suppressMessages(library(MASS)))
suppressWarnings(suppressMessages(library(scales)))
suppressWarnings(suppressMessages(library(showtext)))
suppressWarnings(suppressMessages(library(grid)))
```



# Scatter plot æ•£ç‚¹å›¾

## 1.Scatter plot plus fitted line æ•£ç‚¹å›¾åŠ æ‹Ÿåˆçº¿

```{r Example01, fig.show='asis', fig.width=4, fig.height=2.5}
# è¯»å–æ•°æ®
# Load data
Scatterplot <- read.table("data/figure1e.txt", header = TRUE, sep = "\t", comment.char = "")
colnames(Scatterplot) <- c("PROGENY", "KEGG")

# æ•£ç‚¹å›¾
# Plot
p1 <- ggplot(Scatterplot, aes(x = PROGENY, y = KEGG)) +
  geom_point(size = 3.5, color = "#1F78B4") +  
  geom_smooth(method = "lm", color = "#E31A1C", size = 1.3, linetype = "dashed") + 
  stat_cor(method = "pearson",
           label.sep = '\n',
           p.accuracy = 0.001,
           r.digits = 3,
           label.x = 9.2,
           size = 5, color = "darkred") +
  scale_x_continuous(limits = c(8.9, 10.6), expand = c(0, 0)) +
  scale_y_continuous(limits = c(6.2, 9.8), expand = c(0, 0)) +
  labs(x = "MAPK-PROGENy GES", y = "IL-17 GES") +
  ggtitle("Scatter Plot of PROGENy vs IL-17 GES") + 
  theme_minimal(base_size = 16) +  
  theme(
    axis.text = element_text(color = "black", size = 13),
    axis.title = element_text(size = 15, face = "bold"),
    plot.title = element_text(size = 17, face = "bold", hjust = 0.5),
    panel.border = element_rect(color = "gray", fill = NA, size = 0.8) 
  )

# å°†å›¾å½¢ä¿å­˜ä¸ºPDFæ–‡ä»¶
# Save plot
ggsave("results/scatterplot01.pdf", plot = p1, width = 6, height = 4.5)
```

![](scatterplot01_00.png)


## 2.Scatter plot plus histogram æ•£ç‚¹å›¾åŠ ç›´æ–¹å›¾

å‚è€ƒï¼šhttps://mp.weixin.qq.com/s/3wBXO3gHWj2TpgHqJdBm6w https://mp.weixin.qq.com/s/k6XH4KQaJ6BSDwNO_xUx4w

```{r Example02, fig.show='asis', fig.width=4, fig.height=2.5}
# åˆ›å»ºç¤ºä¾‹æ•°æ®
# Create sample data
covariance <- matrix(c(1, -0.2, -0.2, 1), nrow = 2, byrow = TRUE)
data <- mvrnorm(n = 1000, mu = c(0.5, 3), covariance)
data <- as.data.frame(data)
colnames(data) <- c("Genome size", "GC content")

# æ·»åŠ åˆ†ç»„ä¿¡æ¯
# Add group info
data$group <- "Host unknown"
data$group[order(data$`Genome size`)[c(sample(10:30, 10), sample(500:1000, 300))]] <- "Bacteroidetes"
data$group[order(data$`Genome size`)[c(sample(30:60, 20), sample(200:800, 300))]] <- "Firmicutes"
data$group[order(data$`Genome size`)[c(sample(40:80, 20), sample(100:200, 20))]] <- "Actinobacteria"
data$group[order(data$`Genome size`)[sample(1:1000, 10)]] <- "Proteobacteria"

# è®¾ç½®é¢œè‰²
# Set color
colors <- c("Bacteroidetes" = "#EB746A",  
            "Firmicutes" = "#7AA82C",    
            "Actinobacteria" = "#1EB5B8",
            "Proteobacteria" = "#A07DB7",
            "Host unknown" = "#AAAAAA")  

# æ•£ç‚¹å›¾
# Plot
p21 <- ggplot(data) +
  geom_point(aes(`Genome size`, `GC content`, color = group), alpha = 0.8, size = 3.5) +
  scale_color_manual(values = colors) +
  theme_minimal() +
  theme(
    legend.position = "top",
    legend.title = element_blank(),
    axis.text = element_text(size = 12),
    axis.title = element_text(size = 14, face = "bold"),
    axis.line = element_line(size = 0.6), 
    panel.grid = element_line(color = "grey90")
  ) +
  labs(x = "Genome Size", y = "GC Content")

# ä¸Šæ–¹ç›´æ–¹å›¾
# Top bar plot
p22 <- ggplot(data) +
  geom_histogram(aes(`Genome size`), binwidth = 0.2, fill = "#EB746A", color = "white", size = 0.8) +
  theme_minimal() +
  theme(
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.line = element_line(size = 0.6),   # æ·»åŠ åˆ»åº¦çº¿
    panel.grid = element_blank()
  ) +
  ylab("# of vOTUs")

# å³ä¾§ç›´æ–¹å›¾
# Right bar plot
p23 <- ggplot(data) +
  geom_histogram(aes(`GC content`), binwidth = 0.2, fill = "#A07DB7", color = "white", size = 0.8) +
  theme_minimal() +
  theme(
    axis.title.y = element_blank(),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.line = element_line(size = 0.6),   # æ·»åŠ åˆ»åº¦çº¿
    panel.grid = element_blank()
  ) +
  coord_flip() +
  ylab("# of vOTUs")

# æ‹¼å›¾
# Patchwork
p24 <- ggdraw() + 
  draw_plot(p21, 0.05, 0, .75, .8) +
  draw_plot(p22, 0.05, 0.8, .75, .2) +
  draw_plot(p23, 0.8, 0, .2, .8)

# ä¿å­˜ä¼˜åŒ–åçš„å›¾åƒä¸ºPDF
# Save plot
ggsave("results/scatter_histogram_high.pdf", plot = p24, height = 6, width = 8)
```


![](scatter_histogram_high_00.png)


## 3.Scatter plot plus density plot æ•£ç‚¹å›¾åŠ å¯†åº¦å›¾

```{r Example03, fig.show='asis', fig.width=4, fig.height=2.5}
# æ•°æ®ä¸ºâ€œ2.Scatter plot plus histogram æ•£ç‚¹å›¾åŠ ç›´æ–¹å›¾â€ä½¿ç”¨çš„æ•°æ®
# Data inherted from "2.Scatter plot plus histogram æ•£ç‚¹å›¾åŠ ç›´æ–¹å›¾"

# ç»˜åˆ¶å¯†åº¦å›¾
# Plot density plot
p31 <- ggplot(data, aes(x = `Genome size`, color = group, fill=group)) +
  geom_density(alpha = 0.4, size = 1) +
  scale_y_continuous(expand = c(0,0))+
  scale_color_manual(values = c("#f28c88","#ffbe5d","#9db1c1","#d6d6d6","#77bda1"))+
  scale_fill_manual(values =alpha(c("#f28c88","#ffbe5d","#9db1c1","#d6d6d6","#77bda1"),0.6) )+
  theme_classic()+
  theme(panel.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks = element_line(size=0.6,colour = "black"),
        axis.text.y = element_text(size=10,colour = "black"),
        axis.title.y = element_text(size=16),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
    legend.position = "none"
    )
#p31

p32 <- ggplot(data, aes(x = `GC content`, color = group, fill=group)) +
  geom_density(alpha = 0.4, size = 1) +
  scale_y_continuous(expand = c(0,0))+
  scale_color_manual(values = c("#f28c88","#ffbe5d","#9db1c1","#d6d6d6","#77bda1"))+
  scale_fill_manual(values =alpha(c("#f28c88","#ffbe5d","#9db1c1","#d6d6d6","#77bda1"),0.6) )+
  theme_classic()+
  theme(panel.background = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks = element_line(size=0.6,colour = "black"),
        axis.text.x = element_text(size = 12,colour = "black"),
        axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank(),
    legend.position = "none"
    )+coord_flip()
#p32

# æ‹¼å›¾
# Patchwork
p33 <- ggdraw() + 
  draw_plot_label("a", size = 20)+
  draw_plot(p21, 0.05, 0, .75, .8)+
  draw_plot(p31, 0.05, 0.8, .75, .2)+
  draw_plot(p32, 0.8, 0, .2, .8)
ggsave("results/plot03.pdf", height = 5, width = 8, plot = p33)
#p33
```


![](plot03_00.png)


## 4.Scatter plot plus box plot æ•£ç‚¹å›¾åŠ ç®±çº¿å›¾

```{r Example04, fig.show='asis', fig.width=4, fig.height=2.5}
# è½½å…¥æ•°æ®
# Load data
distance_mat=read.table(paste0("data/OTUID_beta_diversity_yjb_all.txt"), header=T, row.names=1, sep="\t", comment.char="")
distance_mat[1:3, 1:3]
metadata=read.table("data/metadata_double_single3.txt", header=T, row.names=1, sep="\t", comment.char="", stringsAsFactors=F)

df <- metadata
df <- df[rownames(df)%in%rownames(distance_mat), ]

# è¿›è¡ŒPCoAåˆ†æ(PCoA analysis)
# pcoa_result <- cmdscale(distance_mat, eig = TRUE, k = 2)
# pcoa_result2 <- pcoa_result[["points"]]

# bray_dis <- vegdist(varespec, method = 'bray')#åŸºäºBray-curtis è·ç¦»æµ‹ç®—(Based on Bray-Curtis distance measurement)
nmds_dis <- metaMDS(distance_mat, k = 2)#NMDSæ’åºè®¡ç®—ï¼Œä¸€èˆ¬å®šä¹‰ 2 ä¸ªç»´åº¦(NMDS sorting calculation generally defines 2 dimensions)
nmds_dis_site <- data.frame(nmds_dis$points)#è®¡ç®—æ ·ç‚¹å¾—åˆ†(Calculate sample point scores)

df$nmds1 <- nmds_dis_site[, 1]
df$nmds2 <- nmds_dis_site[, 2]

# å®šä¹‰é¢œè‰²
# Define color
colors <- c("Patients" = "#FF6F61", "Healthy" = "#6B5B95")

# ç»˜åˆ¶PCoAæ•£ç‚¹å›¾
# Draw PCoA scatter plot
p41 <- ggplot(df, aes(x = nmds1, y = nmds2, color = Group)) +
  geom_point(alpha = 0.7, size = 3) +
  stat_ellipse(level = 0.68) +
  scale_color_manual(values = colors) +
  theme_bw(base_size = 15) + 
  labs(x = paste0("NMDS1"),
       y = paste0("NMDS2")
       ) +
  theme(
    legend.position = "bottom",
    legend.title = element_text(size = 14, face = "bold"),
    legend.text = element_text(size = 12),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text = element_text(size = 12),
    axis.title = element_text(size = 14, face = "bold"),
    plot.title = element_text(size = 16, face = "bold", hjust = 0.5),
    plot.margin = unit(c(1, 1, 1, 1), "cm")
  )

# æ·»åŠ è¾¹ç¼˜ç®±çº¿å›¾
# Add marginal box plots
p42 <- ggMarginal(p41, type = "boxplot", margins = "both", groupColour = TRUE, groupFill = TRUE)

# æ˜¾ç¤ºå›¾(Show plot)
# print(p)
ggsave("results/NMDS_plot01.pdf", plot = p42, width = 6, height = 6)
```


![](NMDS_plot01_00.png)


## 5.Add error bars and mean to grouped scatter plot åˆ†ç»„æ•£ç‚¹å›¾æ·»åŠ è¯¯å·®æ£’å’Œå¹³å‡å€¼

å‚è€ƒï¼šhttps://mp.weixin.qq.com/s/e6CVI23i-qgWANTaDXAJdw

```{r Example05, fig.show='asis', fig.width=4, fig.height=2.5}
# è½½å…¥æ•°æ®
# Load data
test_design <- read.csv("data/test_design.csv",row.names = 1)
df2 <- test_design[,c(2,5)]

# ç»Ÿè®¡å¹³å‡å€¼
# Calculate the average value
mean_df <- aggregate(df2[,2],by=list(df2[,1]),FUN=mean)
rownames(mean_df) <- mean_df[,1]

# ç»Ÿè®¡æ ‡å‡†å·®
# Standard deviation
sd_df <- aggregate(df2[,2],by=list(df2[,1]),FUN=sd)
rownames(sd_df) <- sd_df[,1]
sd_df <- sd_df[,-1]

# è®¡ç®—æ ‡å‡†è¯¯
# Calculate the standard error
sd_df<-sd_df/2 
se<-as.data.frame(sd_df)

# åˆå¹¶æ•°æ®
# Merge data
df1<-cbind(mean_df,se)
colnames(df1)<-c("group","mean","se")
colnames(df2)<-c("group","Gene_Abundance")

# ç»˜å›¾
# Plot
p51 <- ggplot()+stat_summary(fun = "mean",geom = "crossbar", mapping = aes(x=df1$group,y=df1$mean),
                      width=0.65,size=1.0, color = "gray2")+
  geom_jitter(data=df2,mapping=aes(x=group,y=Gene_Abundance,fill = group,colour = group,shape = group),
              size = 3.5, height = 0.01,width = 0.1)+ 
  scale_color_manual(values = c("#EB746A","#7AA82C","#1EB5B8","#A07DB7"))+ 
  geom_errorbar(data=df1,mapping=aes(x = group,ymin = mean-se, ymax = mean+se), width = 0.2, 
                 color = "gray2",size=1)+
  labs(y="Abundance", x="")+
  theme_bw(base_line_size = 1.05,base_rect_size =1.05)+
  theme(panel.grid.major=element_blank(),panel.grid.minor=element_blank())+
  theme(axis.text=element_text(colour='black',size=12))

# æ”¹å˜æ ·å¼
# Change the style
p52 <- ggplot()+stat_summary(fun = "mean",geom = "crossbar", mapping = aes(x=df1$group,y=df1$mean),
                      width=0.2,size=0.5, height =0.01 ,color = c("#EB746A","#7AA82C","#1EB5B8","#A07DB7"))+
  geom_jitter(data=df2,mapping=aes(x=group,y=Gene_Abundance,fill = group,colour = group),
              size = 3.5,height = 0.01,width = 0.1, shape = 20)+ 
  scale_color_manual(values = c("#EB746A","#7AA82C","#1EB5B8","#A07DB7"))+ 
  geom_errorbar(data=df1,mapping=aes(x = group,ymin = mean-se, ymax = mean+se), width = 0.2, 
                 color = c("#EB746A","#7AA82C","#1EB5B8","#A07DB7"),size=1)+
  labs(y="Abundance", x="")+
  theme_classic()+
  theme(panel.grid.major=element_blank(),panel.grid.minor=element_blank())+
  theme(axis.text=element_text(colour='black',size=12))

# æ”¹å˜æ ·å¼
# Change the style
p53 <- ggplot(df2, aes(group, Gene_Abundance))+
  stat_summary(fun.data="mean_sdl", fun.args = list(mult=1), geom="pointrange", color = "pink",size = 1.2)+
  stat_summary(fun="mean", fun.args = list(mult=1), geom="point", color = "white",size = 4)+
  geom_jitter(data=df2,mapping=aes(x=group,y=Gene_Abundance,fill = group,colour = group),
              size = 3.5,height = 0.01,width = 0.1, shape = 20)+ 
  scale_color_manual(values = c("#EB746A","#7AA82C","#1EB5B8","#A07DB7"))+
  labs(y="Abundance", x="")+
  theme_classic()

# æ·»åŠ æ˜¾è‘—æ€§æ ‡è®°
# Add significance
p54 <- ggplot()+stat_summary(fun = "mean",geom = "crossbar", mapping = aes(x=df1$group,y=df1$mean),
                      width=0.2,size=0.5, height =0.01 ,color = c("#EB746A","#7AA82C","#1EB5B8","#A07DB7"))+
  geom_jitter(data=df2,mapping=aes(x=group,y=Gene_Abundance,fill = group,colour = group),
              size = 3.5,height = 0.01,width = 0.1)+ 
  scale_color_manual(values = c("#EB746A","#7AA82C","#1EB5B8","#A07DB7"))+ 
  geom_errorbar(data=df1,mapping=aes(x = group,ymin = mean-se, ymax = mean+se), width = 0.2, 
                 color = c("#EB746A","#7AA82C","#1EB5B8","#A07DB7"),size=1)+
  labs(y="Abundance", x="")+
  theme_classic()+
  scale_y_continuous(limits = c(0, 50))+
  geom_signif(data = df2, mapping=aes(x=group,y=Gene_Abundance), 
              comparisons = list(c("AA", "BB"), 
                                 c("AA", "CC"),
                                 c("AA", "DD")
                                 ),
              map_signif_level=F, 
              tip_length=c(0,0,0,0,0,0,0,0,0,0,0,0), 
              y_position = c(45,40,35), 
              size=1,
              textsize = 4,
              test = "wilcox.test"
              )
ggsave("results/scatter_plot_error_bar.pdf", p54, width = 6, height = 6)
```


![](scatter_plot_error_bar_00.png)



## 6.Centroid + Error Bar 1 è´¨å¿ƒ+è¯¯å·®æ£’1

æ¡ˆä¾‹ï¼šWang, T., et al. (2024). "Divergent age-associated and metabolism-associated gut microbiome signatures modulate cardiovascular disease risk." Nat Med 30(6): 1722-1731. https://doi.org/10.1038/s41591-024-03038-y

å‚è€ƒï¼šhttps://mp.weixin.qq.com/s/h2EwRvvM7BXVlwpzC7_gNQ


```{r Example06, fig.show='asis', fig.width=4, fig.height=2.5}
# è¯»å–æ•°æ®
# Load data
mydata <- read.csv("data/pca_data.csv", header = TRUE)
data1 <- mydata[, 1:10]
data <- scale(data1)

# è¿›è¡ŒPCAåˆ†æ
# Perform PCA analysis
res.pca <- PCA(data, graph = FALSE)

# æå–æ ·æœ¬PC1ã€PC2åæ ‡å’Œåˆ†ç»„ä¿¡æ¯
# Extract sample PC1, PC2 coordinates and grouping information
result <- data.frame(res.pca$ind$coord)
pc1 <- result[, 1]
pc2 <- result[, 2]
group <- mydata[, 11]

# åˆå¹¶æ•°æ®ç”¨äºç»˜å›¾
# Combining data for plotting
plotdata <- data.frame(pc1 = pc1, pc2 = pc2, group = group)

# è®¡ç®—å‡å€¼ã€æ ‡å‡†å·®ã€æ ‡å‡†è¯¯å·®å’Œ95%ç½®ä¿¡åŒºé—´
# Calculation of mean, standard deviation, standard error and 95% confidence interval
se <- function(x) sd(x) / sqrt(length(x))
conf_95 <- function(x) t.test(x, conf.level = 0.95)$conf.int[2]

sumdata <- summaryBy(pc1 + pc2 ~ group, data = plotdata, FUN = c(mean, sd, se, conf_95))

# ç»˜å›¾
# Plot
p61 <- ggplot(sumdata, aes(x = pc1.mean, y = pc2.mean, color = group)) +
  geom_point(size = 3) +
  scale_color_manual(values = c("#1b9e77", "#d95f02", "#7570b3", "#e7298a", "#66a61e", "#e6ab02")) +
  #theme_minimal() +
  theme_bw()+
  #theme_classic()+
  labs(x = 'PC1', y = 'PC2') +
  theme(
    text = element_text(family = "Arial", size = 16),
    axis.title = element_text(face = "bold"),
    axis.ticks.length = unit(0.2, "cm"),
    axis.text = element_text(color = "black"),
    legend.position = "right",
    legend.title = element_blank()
  )

# æ·»åŠ è¯¯å·®æ£’
# Add error bars
p62 <- p61 +
  geom_errorbarh(aes(xmin = pc1.mean - pc1.sd, xmax = pc1.mean + pc1.sd), height = 0.1, size = 0.8) +
  geom_errorbar(aes(ymin = pc2.mean - pc2.sd, ymax = pc2.mean + pc2.sd), width = 0.1, size = 0.8)

# æ·»åŠ è¯¯å·®æ£’
# Adding error bars
p63 <- p61 +
  geom_errorbarh(aes(xmin = pc1.mean - pc1.se, xmax = pc1.mean + pc1.se), height = 0.1, size = 0.8) +
  geom_errorbar(aes(ymin = pc2.mean - pc2.se, ymax = pc2.mean + pc2.se), width = 0.1, size = 0.8)

# æ·»åŠ è¯¯å·®æ£’
#  Add error bars
showtext_auto()
p64 <- p61 +
  geom_errorbarh(aes(xmin = pc1.mean - pc1.conf_95, xmax = pc1.mean + pc1.conf_95), height = 0.1, size = 0.8) +
  geom_errorbar(aes(ymin = pc2.mean - pc2.conf_95, ymax = pc2.mean + pc2.conf_95), width = 0.1, size = 0.8) +
  theme(text = element_text(family = "Arial", size = 16)) 

# ä¿å­˜å›¾åƒä¸ºPDFæ–‡ä»¶
# Save as PDF
ggsave("results/Centroid_Error_Bar.pdf", plot = p64, width = 7, height = 4)
```


![](Centroid_error_bar_00.png)



## 7.Centroid + Error Bar 2 è´¨å¿ƒ+è¯¯å·®æ£’2

å‚è€ƒï¼šhttps://mp.weixin.qq.com/s/7SHgJXhctRCi4QnHyaAZug

æ¡ˆä¾‹ï¼šPÃ¤rn, J., Verhoeven, J.T.A., Butterbach-Bahl, K. et al. Nitrogen-rich organic soils under warm well-drained conditions are global nitrous oxide emission hotspots. Nat Commun 9, 1135 (2018). https://doi.org/10.1038/s41467-018-03540-1


```{r Example07, fig.show='asis', fig.width=4, fig.height=2.5}
# åŠ è½½æ•°æ®
# Load data
data1 <- read.csv("data/test_data2.csv",header = T)

# é‡æ–°è®¡ç®—è´¨å¿ƒæ•°æ®
# Recalculate the centre of mass data
X <- aggregate(X ~ group, data1, function(x) c(mean = mean(x), sd = sd(x)))
Y <- aggregate(Y ~ group, data1, function(x) c(mean = mean(x), sd = sd(x)))
data2 <- data.frame(
  group=X$group,
  X=X$X[,1],
  X_error=X$X[,2],
  Y=Y$Y[,1],
  Y_error=Y$Y[,2]
)

# ç»˜å›¾
# Plot
p71 <- ggplot(data2, aes(X, Y)) +
  geom_errorbar(aes(xmin = X - X_error, xmax = X + X_error, color = group), 
                linewidth = 0.6, width = 0.2, alpha = 0.7) +
  geom_errorbar(aes(ymin = Y - Y_error, ymax = Y + Y_error, color = group), 
                linewidth = 0.6, width = 0.2, alpha = 0.7) +
  geom_point(aes(color = group), size = 3, alpha = 0.8) +
  scale_color_manual(values = c("#D55E00", "#0072B2", "#F0E442", "#009E73",
                                "#CC79A7", "#56B4E9", "#E69F00", "#999999",
                                "#1A85FF", "#FF6666", "#66CC99", "#006400",
                                "#8B0000", "#DAA520", "#556B2F", "#FF4500",
                                "#4682B4", "#9400D3")) +
  geom_smooth(method = "lm", color = "darkblue", fill = "#ADD8E6", se = TRUE, 
              formula = y ~ x, linetype = "dashed", alpha = 0.3) +
  stat_poly_eq(formula = y ~ x, 
               aes(label = paste(after_stat(eq.label), after_stat(adj.rr.label), sep = "~~~")), 
               parse = TRUE, size = 4, label.x.npc = "right", label.y.npc = "top") +
  labs(x = " category_A", y = " category_B", color = "group") +
  #theme_minimal(base_size = 15) +
  theme_bw()+
  theme(
    legend.position = "right",        
    legend.title = element_text(size = 12, face = "bold"),
    legend.text = element_text(size = 10), 
    axis.title = element_text(size = 16),   
    axis.text = element_text(size = 14),    
    panel.grid.major = element_line(size = 0.5, linetype = "dotted", color = "gray80"),  
    panel.grid.minor = element_blank()      
  )

# ä¿å­˜PDF
# Save as PDF
ggsave("results/Centroid_Error_Bar02.pdf", plot = p71, width = 8, height = 6, dpi = 300)
```


![](Centroid_Error_Bar02_00.png)


If used this script, please cited:
ä½¿ç”¨æ­¤è„šæœ¬ï¼Œè¯·å¼•ç”¨ä¸‹æ–‡ï¼š

**Yong-Xin Liu**, Lei Chen, Tengfei Ma, Xiaofang Li, Maosheng Zheng, Xin Zhou, Liang Chen, Xubo Qian, Jiao Xi, Hongye Lu, Huiluo Cao, Xiaoya Ma, Bian Bian, Pengfan Zhang, Jiqiu Wu, Ren-You Gan, Baolei Jia, Linyang Sun, Zhicheng Ju, Yunyun Gao, **Tao Wen**, **Tong Chen**. 2023. EasyAmplicon: An easy-to-use, open-source, reproducible, and community-based pipeline for amplicon data analysis in microbiome research. **iMeta** 2: e83. https://doi.org/10.1002/imt2.83

Copyright 2016-2024 Defeng Bai <baidefeng@caas.cn>, Chuang Ma <22720765@stu.ahau.edu.cn>, Jiani Xun <15231572937@163.com>, Yong-Xin Liu <liuyongxin@caas.cn>
