---
title: "计数值(count)特征表差异比较及可视化-火山图、曼哈顿图和热图"
author: "Yong-Xin Liu"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 设置输入输出文件和参数

修改下面`default=`后面的文件和参数。

输入文件为原始OTU表(otutab.txt)+物种注释(taxonomy.txt)+分组信息(metadata.txt)

输入文件"-i", "--input"，OTU table in counts; 原始OTU表counts值；

实验设计"-d", "--design"，默认`metadata.txt`，可手动修改文件位置；

物种注释"-t", "--taxonomy"，Taxonomy file; 物种注释

分组列名"-n", "--group"，默认将metadata.txt中的group列作为分组信息，可修改为任意列名；

输入文件前缀"-o", "--output"，默认为空时，输出为当前目录前缀为KO-WT_all/sig.txt统计表格，A-B_volcano/manhattan/heatmap.pdf组比较图片。

物种图例顺序"-T", "--top10tax"，Top 10 phylum; 自定义门图例

比较组"-c", "--compare"，Groups comparison; 组间比较，默认为KO-WT

Pvalue阈值"-p", "--pvalue"，Threshold of P-value, 显著性阈值

假阳性率阈值"-f", "--fdr"，Threshold of FDR, 假阳性率阈值

图片宽"-w", "--width"，默认8英寸，根据图像布局可适当增大或缩小

图片高"-e", "--height"，默认5英寸，根据图像布局可适当增大或缩小

```{r main}

# 1.2 解析命令行
# 设置清华源加速下载
site="https://mirrors.tuna.tsinghua.edu.cn/CRAN"
# 判断命令行解析是否安装，安装并加载
if (!suppressWarnings(suppressMessages(require("optparse", character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)))) {
  install.packages(p, repos=site)
  require("optparse",character.only=T) 
}
# 解析参数-h显示帮助信息
if (TRUE){
  option_list = list(
    make_option(c("-i", "--input"), type="character", default="otutab.txt",
                help="OTU table in counts; 原始OTU表counts值 [default %default]"),
    make_option(c("-d", "--design"), type="character", default="metadata.txt",
                help="Design file; 实验设计文件 [default %default]"),
    make_option(c("-t", "--taxonomy"), type="character", default="taxonomy.txt",
                help="Taxonomy file; 物种注释 [default %default]"),
    make_option(c("-T", "--top10tax"), type="character", default="tax_phylum.top10",
                help="Top 10 phylum; 自定义门图例 [default %default]"),    
    make_option(c("-n", "--group"), type="character", default="Group",
                help="Group name; 分组列名 [default %default]"),
    make_option(c("-c", "--compare"), type="character", default="KO-OE",
                help="Groups comparison; 组间比较 [default %default]"),
    make_option(c("-p", "--pvalue"), type="numeric", default=0.05,
                help="Threshold of P-value, 显著性阈值 [default %default]"),
    make_option(c("-f", "--fdr"), type="numeric", default=0.1,
                help="Threshold of FDR, 假阳性率阈值 [default %default]"),
    make_option(c("-o", "--output"), type="character", default="",
                help="Output prefix; 结果前缀.txt表/pdf图 [default %default]"),
    make_option(c("-w", "--width"), type="numeric", default=8,
                help="Figure width; 图片宽 [default %default]"),
    make_option(c("-e", "--height"), type="numeric", default=5,
                help="Figure heidth; 图片高 [default %default]")
  )
  opts = parse_args(OptionParser(option_list=option_list))
  
  # 调置如果无调设置输出，根据其它参数设置默认输出
  if (opts$output==""){
    opts$output=paste("./",opts$compare, sep = "")}
  
  # 显示输入输出参数，用户确认是否正确
  print("Parameters are as follows. Please check it!")
  print(paste("The input data matrix file is ", opts$input,  sep = ""))
  print(paste("The design file is ", opts$design,  sep = ""))
  print(paste("The taxonomy file is ", opts$taxonomy,  sep = ""))
  print(paste("Top 10 phylum file is ", opts$top10tax,  sep = ""))
  print(paste("Group name is ", opts$group,  sep = ""))
  print(paste("Group compare is ", opts$compare,  sep = ""))
  print(paste("Threshold of P-value is ", opts$pvalue,  sep = ""))
  print(paste("Threshold of FDR is ", opts$fdr,  sep = ""))
  print(paste("Output figure width ", opts$width,  sep = ""))
  print(paste("Output figure height ", opts$height,  sep = ""))
  print(paste("The output file is ", opts$output, sep = ""))
  print("",quote = F)
}



# 2. 依赖关系检查、安装和加载

# 2.1 安装CRAN来源常用包
# 依赖包列表：差异分析、绘图、热图、数据变换和开发者工具
package_list = c("ggplot2","pheatmap","dplyr","devtools")
# 判断R包加载是否成功来决定是否安装后再加载
for(p in package_list){
  if(!suppressWarnings(suppressMessages(require(p, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)))){
    install.packages(p, repos=site)
    suppressWarnings(suppressMessages(library(p, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)))
  }
}

# 2.2 安装bioconductor常用包
# 基于reads counts值组间差异分析包
package_list = c("limma","edgeR")
for(p in package_list){
  if(!suppressWarnings(suppressMessages(require(p, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)))){
    source("https://bioconductor.org/biocLite.R")
    biocLite(p)
    suppressWarnings(suppressMessages(library(p, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)))
  }
}

# # 2.3 安装Github常用包
# # ggplot2美化
# package_list = c("kassambara/ggpubr")
# for(p in package_list){
#   q=unlist(strsplit(p,split = "/"))[2]
#   if(!suppressWarnings(suppressMessages(require(q, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)))){
#     install_github(p)
#     suppressWarnings(suppressMessages(library(q, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)))
#   }
# }



# 3. 读取输入文件

# 读取OTU表
dat = read.table(opts$input, header=T, row.names= 1, sep="\t", comment.char = "") 

# 读取实验设计
design = read.table(opts$design, header=T, row.names= 1, sep="\t", comment.char = "") 

# 将选定的分组列统一命名为group
design$group=design[,opts$group]



# 4. 统计

# edgeR函数
run_edgeR = function(dat,design,compare){
  # 筛选比较组
  group_list=strsplit(opts$compare,'-')[[1]]
  idx = design$group %in% group_list
  sub_design=design[idx,]
  sub_dat=as.matrix(dat[,rownames(sub_design)])  

  d = DGEList(counts=sub_dat,group=factor(sub_design$group))
  d = calcNormFactors(d)
  # check samples is in right groups
  d$samples 

  # design.mat = model.matrix(~factor(sub_design$group))
  design.mat = model.matrix(~ 0 + factor(sub_design$group))
  rownames(design.mat)=colnames(sub_dat)
  colnames(design.mat)=levels(factor(sub_design$group))
  DAO = estimateDisp(d,design.mat)
  fit = glmFit(DAO,design.mat)
  BvsA <- makeContrasts(contrasts = opts$compare, levels=design.mat)
  lrt = glmLRT(fit,contrast=BvsA)
  # lrt = glmLRT(fit,coef=2)
  
  nrDAO=as.data.frame(topTags(lrt, n=nrow(dat)))
  nrDAO=as.data.frame(nrDAO)
  head(nrDAO)

  nrDAO$logFC=round(nrDAO$logFC,3)
  nrDAO$logCPM=round(nrDAO$logCPM,3)
  nrDAO$level = ifelse(nrDAO$logFC>0 & nrDAO$PValue<opts$pvalue & nrDAO$FDR < opts$fdr, "Enriched",ifelse(nrDAO$logFC<0 & nrDAO$PValue<opts$pvalue & nrDAO$FDR < opts$fdr, "Depleted","NotSig"))
  nrDAO$level=factor(nrDAO$level,levels = c("Enriched","Depleted","NotSig"))
  
  # 如果存在物种注释，添加注释信息
  if (file.exists(opts$taxonomy)){
  tax = read.table(opts$taxonomy, header=T, row.names= 1, sep="\t", comment.char = "") 
  tax = tax[rownames(nrDAO),]
  nrDAO=cbind(nrDAO,tax)
  
  # Draw manhattan plot and color by phylum
  x=nrDAO
  x$otu=rownames(x)
  x$neglogp=-log10(x$PValue)
  # order taxonomy
  x = arrange(x, Phylum, Class, Order, Family, Genus,otu)
  rownames(x) = x$otu
  
  # Taxonomy top 10, other in low abundance
  x$tax=gsub("p__","",x$Phylum)
  if (file.exists(opts$top10)){
    top10 = read.table(opts$top10tax)
    top10 = as.vector(top10$V1)
  }else{
    top10=sort(c("Acidobacteria","Actinobacteria","Bacteroidetes",
            "Chloroflexi","Cyanobacteria","Firmicutes","Proteobacteria",
            "Verrucomicrobia","Planctomycetes"))
  }
  print(paste("The top",length(top10)[1],"phylum as legends.", sep=" "))
  print(top10)
  print("",quote = F)
  x$tax=as.vector(x$tax)
  # levels(x$tax)=c(unique(x$tax),"Low Abundance")
  if (length(unique(x$tax)) > length(top10)){
    x[!(x$tax %in% top10),]$tax = "Low Abundance"
  }
  x$otu=factor(x$otu,levels = x$otu)
  FDR = min(x$neglogp[x$level=="Enriched"])
  x$Level=x$level
  x$Phylum=x$tax
  p = ggplot(x, aes(x=otu, y=neglogp, color=Phylum, size=logCPM, shape=Level)) +
    geom_point(alpha=.7) + 
    geom_hline(yintercept=FDR, linetype=2, color="lightgrey") +
    scale_shape_manual(values=c(17, 25, 20))+
    scale_size(breaks=c(5, 10, 15)) +
    labs(x="OTUs", y="-log10(P)", title=paste(group_list[1], "vs", group_list[2], sep=" ")) +
    theme_classic() +
    theme(axis.ticks.x=element_blank(),axis.text.x=element_blank(),legend.position="right")
  p
  ggsave(paste(opts$output, "_manhattan.pdf", sep=""), p, 
         width = opts$width*2, height = opts$height)
  }
  
  # Add MeanA and MeanB in percentage
  # normlization to percentage
  norm = t(t(sub_dat)/colSums(sub_dat,na=T))*100
  # check norm is right?
  colSums(norm)
  # calculate groupA mean
  A_list = subset(sub_design, group %in% group_list[1])
  A_norm = norm[, rownames(A_list)]
  A_mean = as.data.frame(rowMeans(A_norm))
  colnames(A_mean)=c("MeanA")
  # calculate groupB mean
  B_list = subset(sub_design, group %in% group_list[2])
  B_norm = norm[, rownames(B_list)]
  B_mean = as.data.frame(rowMeans(B_norm))
  colnames(B_mean)=c("MeanB")
  # merge and reorder
  Mean = round(cbind(A_mean, B_mean, A_norm, B_norm),3)
  Mean = Mean[rownames(nrDAO),]   
  output=cbind(nrDAO[,-3],Mean)

  # write all OTU for volcano plot and manhattan plot
  write.table("OTUID\t", file=paste(opts$output,"_all.txt",sep=""),append = F, quote = F, eol = "", row.names = F, col.names = F)
  write.table(output,file=paste0(opts$output,"_all.txt",sep=""),append = T,quote = F,sep = '\t',row.names = T)

  # 计算上、下调OTUs数量
  NoE= dim(output[output$level=="Enriched",])[1]
  NoD= dim(output[output$level=="Depleted",])[1]
  # 绘制火山图，
  p = ggplot(output, aes(x=logFC, y=logCPM, color=level)) + 
    geom_point() + xlim(-4, 4) + theme_classic()+
    scale_colour_manual(values=c("red","green","grey")) + 
    labs(x="log2(fold change)", y="log2(count per million)", 
         title=paste(group_list[1], "vs", group_list[2], sep=" "))+ 
    annotate("text",x=-3,y=15,label=paste(NoD,sep=""))+ 
    annotate("text",x=3,y=15,label=paste(NoE,sep=""))
  p
  ggsave(paste(opts$output, "_volcano.pdf", sep=""), p, 
         width = opts$width, height = opts$height)

  # 数据筛选，pvalue < 0.05，FDR < 0.2
  output=output[output$PValue < opts$pvalue,]
  output=output[output$FDR < opts$fdr,]
  # 保存筛选结果于sig.txt结尾文件中
  write.table("OTUID\t", file=paste(opts$output,"_sig.txt",sep=""),append = F, quote = F, eol = "", row.names = F, col.names = F)
  write.table(output,file=paste0(opts$output,"_sig.txt",sep=""),append = T,quote = F,sep = '\t',row.names = T)

  
# 保存三元图enriched结果
  g=group_list
    write.table("OTUID\t", file=paste0("otu_",g[1], "vs", g[2],"_enriched.txt"),append = F, quote = F, eol = "", row.names = F, col.names = F)
  write.table(output[output$level == "Enriched",],file=paste0("otu_",g[1], "vs", g[2],"_enriched.txt"),append = T,quote = F,sep = '\t',row.names = T)

      write.table("OTUID\t", file=paste0("otu_",g[2], "vs", g[1],"_enriched.txt"),append = F, quote = F, eol = "", row.names = F, col.names = F)
  write.table(output[output$level == "Depleted",],file=paste0("otu_",g[2], "vs", g[1],"_enriched.txt"),append = T,quote = F,sep = '\t',row.names = T)
    
  if (file.exists(opts$taxonomy)){
  # 绘差异OTUs有分组、物种门的热图pheatmap
    
  # 制作注释行变化类型和分类学门水平的数据框
  anno_row=data.frame(Level = output$level, 
                            Taxonomy=output$Phylum, 
                            row.names = rownames(output))
  # 制作注释列分组信息
  anno_col=data.frame(Group = sub_design$group,
                      row.names = rownames(sub_design))
  
  # 绘制热图
  pheatmap(norm[rownames(output),],
           scale = "row",
           cutree_rows=2,cutree_cols = 2,
           annotation_col = anno_col, annotation_row = anno_row,
           filename = paste(opts$output, "_heatmap.pdf", sep=""),
           width=opts$width, height=opts$height, 
           annotation_names_row= T,annotation_names_col=F,
           show_rownames=F,show_colnames=T,
           main = paste("Differential abundance OTUs of",group_list[1], "vs", group_list[2],sep=" "),
           fontsize=7,display_numbers=F)
}
}

# edgeR计算组间差异
run_edgeR(dat,design,opts$compare)

# 提示工作完成
print(paste("Output in files in ", opts$output,"*.txt/pdf. All works done!!!", sep = ""))
print("",quote = F)
```
使用此脚本，请引用下文：

If used this script, please cited:

**Yong-Xin Liu**, Yuan Qin, **Tong Chen**, et. al. A practical guide to amplicon and metagenomic analysis of microbiome data. ***Protein Cell*** 41, 1-16, doi: [10.1007/s13238-020-00724-8](https://doi.org/10.1007/s13238-020-00724-8) (2020).

Jingying Zhang, **Yong-Xin Liu**, Na Zhang, et. al. NRT1.1B is associated with root microbiota composition and nitrogen use in field-grown rice. ***Nature Biotechnology*** 37, 676-684, doi: [10.1038/s41587-019-0104-4](https://doi.org/10.1038/s41587-019-0104-4) (2019).

Copyright 2016-2020 Yong-Xin Liu <metagenome@126.com>