---
title: "生成物种分类汇总——LEfSE输入文件"
author: "Yong-Xin Liu"
date: '`r Sys.Date()`'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(amplicon)
```

## 读取输入文件

```{r read, warning=FALSE}
# 1. 读取OTU表
otutab = read.table("otutab.txt", header=T, row.names= 1, sep="\t", comment.char = "", stringsAsFactors = F)
# 2. 读取物种注释
tax = read.table("taxonomy.txt", header=T, row.names= 1, sep="\t",comment.char = "", stringsAsFactors = F) 
# 3. 读取样本分组信息
metadata = read.table("metadata.txt", header=T, row.names=1, sep="\t", comment.char="")
# OTU丰度筛选阈值，默认0.1%，主要为圈图展示合适数据的OTU
thre = 0.1
# 样本分组列名
groupID = "Group"
# 文件名前缀
output = "LEfSe"

# 生成各分类级汇总特征表
suppressWarnings(suppressMessages(library(amplicon)))
# 输出标准格式lefse输入文件
format2lefse(otutab, tax, metadata, thre, groupID, paste0(output,".txt"))
# 输出带有分类注释的lefse
format2lefse2(otutab, tax, metadata, thre, groupID, paste0(output,"2.txt"))
# 在当前目录生成文件LEfSe.txt
```
使用此脚本，请引用下文：

If used this script, please cited:

**Yong-Xin Liu**, Yuan Qin, **Tong Chen**, et. al. A practical guide to amplicon and metagenomic analysis of microbiome data. ***Protein Cell*** 41, 1-16, doi: [10.1007/s13238-020-00724-8](https://doi.org/10.1007/s13238-020-00724-8) (2020).

Jingying Zhang, **Yong-Xin Liu**, Na Zhang, et. al. NRT1.1B is associated with root microbiota composition and nitrogen use in field-grown rice. ***Nature Biotechnology*** 37, 676-684, doi: [10.1038/s41587-019-0104-4](https://doi.org/10.1038/s41587-019-0104-4) (2019).

Copyright 2016-2020 Yong-Xin Liu <metagenome@126.com>